### PR TITLE
163 account recovery code

### LINKED ISSUE #163: Account recovery (backup) mfa codes
We should support having backup mfa codes so that single-device accounts can be restored.

### PR DESCRIPTION
**Pull request by victorcwai**
Fixes #163 (WIP) 
Looking for feedback!
Draft code for using backup code to login and how to create it. I mostly follow the code for doing authentication with totp/password generation.

- [x] cargo fmt has been run
- [ ] cargo clippy has been run
- [x] cargo test has been run and passes
- [ ] book chapter included (if relevant)
- [ ] design document included (if relevant)

### CODE REVIEW

File: kanidmd/src/lib/actors/v1_write.rs [hunk 1]
```diff
@@ -624,6 +625,26 @@ impl QueryServerWriteV1 {
...
+                                audit,
+                                "Failed to begin internal_credential_set_message: {:?}",
+                                e
+                            );
+                            e
+                        })?;
+                        idms_prox_write
+                            .generate_backup_code(&mut audit, &gbe)
+                            .and_then(|r| idms_prox_write.commit(&mut audit).map(|_| r))
+                            .map(SetCredentialResponse::Token)
```
Comment (Firstyear) [inline kanidmd/src/lib/actors/v1_write.rs:646]: This should probably use a new type like BackupCodes on the `SetCredentialResponse` enum so it's really clear what these are.

File: kanidmd/src/lib/core/https.rs [hunk 2]
```diff
@@ -1278,6 +1307,11 @@ pub fn create_https_server(
     account_route
         .at("/:id/_credential/:cid/_lock")
         .get(do_nothing);
+    account_route
+        .at("/:id/_credential/backup_code")
```
Comment (Firstyear) [inline kanidmd/src/lib/core/https.rs:1311]: These should be under :cid or `.at("/:id/_credential/primary")`. This is for future multiple credential handling

File: kanidmd/src/lib/credential/mod.rs [hunk 3]
```diff
@@ -211,6 +213,34 @@ impl Password {
     }
 }
 
+#[derive(Clone, Debug)]
+pub struct BackupCode {
```
Comment (Firstyear) [inline kanidmd/src/lib/credential/mod.rs:217]: Shouldn't this be BackupCodes (plural, append an S here).

File: kanidmd/src/lib/credential/mod.rs [hunk 4]
```diff
@@ -706,6 +768,59 @@ impl Credential {
         }
     }
 
+    pub(crate) fn update_backup_code(&self, backup_code: HashSet<String>) -> Self {
+        let type_ = match &self.type_ {
+            // TODO: we should panic/reject CredentialType != PasswordMfa?
```
Comment (Firstyear) [inline kanidmd/src/lib/credential/mod.rs:773]: we should reject (not panic), so this would change to `Result<Self, OperationError>` to represent that errorstate

File: kanidmd/src/lib/credential/mod.rs [hunk 5]
```diff
@@ -706,6 +768,59 @@ impl Credential {
...
+            CredentialType::Password(pw) | CredentialType::GeneratedPassword(pw) => {
+                debug_assert!(false);
+                CredentialType::PasswordMfa(
```
Comment (Firstyear) [inline kanidmd/src/lib/credential/mod.rs:776]: I don't think this is valid to do this convert, since a backup code exists to recover from lost webauthn/totp.

File: kanidmd/src/lib/credential/mod.rs [hunk 6]
```diff
@@ -706,6 +768,59 @@ impl Credential {
...
+            CredentialType::PasswordMfa(pw, totp, wan, _) => CredentialType::PasswordMfa(
+                pw.clone(),
+                totp.clone(),
+                wan.clone(),
+                Some(BackupCode {
+                    code_set: backup_code,
+                }),
+            ),
+            CredentialType::Webauthn(wan) => {
+                debug_assert!(false);
```
Comment (Firstyear) [inline kanidmd/src/lib/credential/mod.rs:794]: Should return a Err() instead.

File: kanidmd/src/lib/credential/mod.rs [hunk 7]
```diff
@@ -706,6 +768,59 @@ impl Credential {
...
+            }
+        };
+        Credential {
+            type_,
+            claims: self.claims.clone(),
+            uuid: self.uuid,
+        }
+    }
+
+    // pub(crate) fn remove_backup_code(&self) -> Self {
```
Comment (Firstyear) [inline kanidmd/src/lib/credential/mod.rs:805]: I think you should uncomment this, it's good :) 

If you make it so only PasswordMfa takes backup code, it's a much simpler code path. You could probabry consider returning Err() when not PMFA here though so that we respond that the transform is invalid though.

File: kanidmd/src/lib/credential/mod.rs [hunk 8]
```diff
@@ -731,8 +846,8 @@ impl CredentialType {
     fn is_valid(&self) -> bool {
         match self {
             CredentialType::Password(_) | CredentialType::GeneratedPassword(_) => true,
-            CredentialType::PasswordMfa(_, m_totp, webauthn) => {
-                m_totp.is_some() || !webauthn.is_empty()
+            CredentialType::PasswordMfa(_, m_totp, webauthn, _) => {
+                m_totp.is_some() || !webauthn.is_empty() // ignore backup code (it should only be a complement for totp/webauth)
```
Comment (Firstyear) [inline kanidmd/src/lib/credential/mod.rs:850]: I think this check won't be needed if you change the code above to only set backup code if totp/webauthn are some. (which they have to be if you are password MFA anyway).

File: kanidmd/src/lib/idm/account.rs [hunk 9]
```diff
@@ -357,6 +358,39 @@ impl Account {
         }
     }
 
+    pub(crate) fn gen_backup_code_mod(
+        &self,
+        backup_code: HashSet<String>,
```
Comment (Firstyear) [inline kanidmd/src/lib/idm/account.rs:363]: plural

File: kanidmd/src/lib/idm/authsession.rs [hunk 10]
```diff
@@ -301,6 +309,23 @@ impl CredHandler {
                             CredState::Denied(BAD_TOTP_MSG)
                         }
                     }
+                    (AuthCredential::BackupCode(code_chal), _, _, Some(backup_code)) => {
```
Comment (Firstyear) [inline kanidmd/src/lib/idm/authsession.rs:312]: You need to extend and test this state machine flow in the tests at the bottom of this file. This is a security critical code path, so requires *extensive* testing. However, the logic here appears to be correct IMO.

File: kanidmd/src/lib/idm/server.rs [hunk 11]
```diff
@@ -1356,6 +1359,46 @@ impl<'a> IdmServerProxyWriteTransaction<'a> {
...
+        &mut self,
+        au: &mut AuditScope,
+        gbe: &GenerateBackupCodeEvent,
+    ) -> Result<String, OperationError> {
+        let account = self.target_to_account(au, &gbe.target)?;
+
+        // Generate a new set of backup code.
+        let s = backup_code_from_random();
+        // TODO: into_iter() consumes s, find a way to get the joined string without cloning?
+        let s_str = s.clone().into_iter().collect::<Vec<String>>().join(", ");
```
Comment (Firstyear) [inline kanidmd/src/lib/idm/server.rs:1373]: Why not just iter() and .join?

File: kanidmd/src/lib/idm/server.rs [hunk 12]
```diff
@@ -1356,6 +1359,46 @@ impl<'a> IdmServerProxyWriteTransaction<'a> {
             })
     }
 
+    /// Generate a new set of backup code and remove the old ones.
+    pub fn generate_backup_code(
+        &mut self,
+        au: &mut AuditScope,
+        gbe: &GenerateBackupCodeEvent,
+    ) -> Result<String, OperationError> {
```
Comment (Firstyear) [inline kanidmd/src/lib/idm/server.rs:1367]: Shouldn't we return a Vec<String> here to the client, and it's up to the front end utility to format this? Given this is structured data after all, and we don't want the client to have to str split on, when they can just process the json which is structured.

File: kanidmd/src/lib/utils.rs [hunk 13]
```diff
@@ -37,6 +38,14 @@ pub fn password_from_random() -> String {
     rand_string
 }
 
+pub fn backup_code_from_random() -> HashSet<String> {
+    let mut set = std::collections::HashSet::with_capacity(8);
```
Comment (Firstyear) [inline kanidmd/src/lib/utils.rs:42]: Minor, but can't you do:

```
(1..8).iter().map(|_| {
    readable_password_from_random()
})
.collect()
```

There is some type-hint magic in rust that pre-hints the capacity to the hashset during the collect so you don't need to do it manually.

File: kanidmd/src/lib/idm/authsession.rs [hunk 14]
```diff
@@ -301,6 +309,23 @@ impl CredHandler {
...
+                        if backup_code.verify(&code_chal) {
+                            pw_mfa.mfa_state = CredVerifyState::Success;
+                            lsecurity!(
+                                au,
+                                "Handler::PasswordMfa -> Result::Continue - BackupCode OK, password -"
+                            );
+                            CredState::Continue(vec![AuthAllowed::Password])
```
Comment (Firstyear) [inline kanidmd/src/lib/idm/authsession.rs:319]: Consider here adding a `DelayedAction` from kanidmd/src/lib/idm/delayed.rs which will remove code String from account with uuid. See "WebauthnCounterIncrement` for inspiration.

File: kanidmd/src/lib/idm/server.rs (hunk 12 above)
Comment (Firstyear) [reply kanidmd/src/lib/idm/server.rs:1367]: Also consider adding a test into `kanidm_client/tests/proto_v1_test.rs` which actually tests these api interactions to create and retrieve the codes, then to use them from an external client perspective.

File: kanidmd/src/lib/idm/account.rs [hunk 15]
```diff
@@ -357,6 +358,39 @@ impl Account {
...
+    // pub(crate) fn gen_backup_code_remove_mod(&self) -> Result<ModifyList<ModifyInvalid>, OperationError> {
+    //     match &self.primary {
+    //         // Change the cred
+    //         Some(primary) => {
+    //             let ncred = primary.remove_totp();
+    //             let vcred = Value::new_credential("primary", ncred);
+    //             Ok(ModifyList::new_purge_and_set("primary_credential", vcred))
+    //         }
+    //         None => {
+    //             // No credential exists, we can't remove what is not real.
```
Comment (Firstyear) [inline kanidmd/src/lib/idm/account.rs:388]: You can remove anything if you try hard enough :)

**Review by Firstyear — COMMENTED** [review]
Looking really good mate!

File: kanidm_proto/src/v1.rs [hunk 16]
```diff
@@ -310,7 +310,7 @@ pub enum CredentialDetailType {
     GeneratedPassword,
     Webauthn(Vec<String>),
     /// totp, webauthn
-    PasswordMfa(bool, Vec<String>),
+    PasswordMfa(bool, Vec<String>, bool),
```
Comment (Firstyear) [inline kanidm_proto/src/v1.rs:313]: You may want this to be a number of how many codes are remaining so they know they may need to reset it soon if they are low.

File: kanidmd/src/lib/credential/mod.rs [hunk 17]
```diff
@@ -214,30 +214,33 @@ impl Password {
 }
 
 #[derive(Clone, Debug)]
-pub struct BackupCode {
-    // TODO: any other attributes?
+pub struct BackupCodes {
+    // WIP_TODO: any other attributes?
```
Comment (Firstyear) [inline kanidmd/src/lib/credential/mod.rs:218]: For now this is okay :) but still good for it to be in a struct incase we add moe.

File: kanidmd/src/lib/credential/mod.rs [hunk 18]
```diff
@@ -214,30 +214,33 @@ impl Password {
...
     code_set: HashSet<String>,
 }
 
-//TODO: No need a separate DbBackupCodeV1 struct probably?
-impl TryFrom<DbBackupCodeV1> for BackupCode {
+//WIP_TODO: No need a separate DbBackupCodeV1 struct probably?
```
Comment (Firstyear) [inline kanidmd/src/lib/credential/mod.rs:222]: No I think we should still have it as seperate incase we extend or change the type in the future.

File: kanidmd/src/lib/credential/mod.rs [hunk 19]
```diff
@@ -768,58 +773,40 @@ impl Credential {
...
 
-    pub(crate) fn update_backup_code(&self, backup_code: HashSet<String>) -> Self {
-        let type_ = match &self.type_ {
-            // TODO: we should panic/reject CredentialType != PasswordMfa?
-            CredentialType::Password(pw) | CredentialType::GeneratedPassword(pw) => {
-                debug_assert!(false);
-                CredentialType::PasswordMfa(
+    pub(crate) fn update_backup_code(
+        &self,
+        backup_codes: BackupCodes, // Should I be passing reference??
```
Comment (Firstyear) [inline kanidmd/src/lib/credential/mod.rs:778]: No, because this function will take ownership of the backup_codes so it's correct to NOT use a reference here (if you used a reference you'd need to .clone() )

File: kanidmd/src/lib/idm/account.rs [hunk 20]
```diff
@@ -360,14 +360,19 @@ impl Account {
 
     pub(crate) fn gen_backup_code_mod(
         &self,
-        backup_code: HashSet<String>,
+        backup_codes: BackupCodes, // Should I be passing reference??
```
Comment (Firstyear) [inline kanidmd/src/lib/idm/account.rs:363]: See above :)

File: kanidmd/src/lib/idm/authsession.rs [hunk 21]
```diff
@@ -309,8 +310,22 @@ impl CredHandler {
...
-                        if backup_code.verify(&code_chal) {
+                    (AuthCredential::BackupCode(code_chal), _, _, Some(backup_codes)) => {
+                        // Clone backup_codes because it's not mutable. If I make it a mutable ref,
+                        // then I have to pass it as &BackupCodes and specify lifetime on BackupCodeRemoval
+                        let mut cloned_codes = backup_codes.clone();
+                        if cloned_codes.verify(&code_chal) {
+                            if let Err(_e) =
+                                async_tx.send(DelayedAction::BackupCodeRemoval(BackupCodeRemoval {
+                                    target_uuid: who,
+                                    updated_codes: cloned_codes,
```
Comment (Firstyear) [inline kanidmd/src/lib/idm/authsession.rs:321]: There is a problem here. Imagine we have two logins, and they each use a seperate backup code, A and B. We'll have two delayed actions queued. One set where A is NOT present, one where B is not present. So we'll "lose" one of the code removals. 

Also imagine the person reset's their backup codes, before this delayed action is processed. We would then process his and "reset back" all their codes! 

The correct action here is not to clone the codes to send in the delayed action, but to send the *code that was used*. and then the delayed action looks at the set within a write txn and removes that code, if present, or if not present, skips (because the codes were reset). 

This avoids your cloning issue too :) 

tl;dr - 
```
BackupCodeRemoval {
    target_uuid: Uuid,
    code_to_remove: String
}
```

File: kanidmd/src/lib/utils.rs [hunk 22]
```diff
@@ -39,11 +39,10 @@ pub fn password_from_random() -> String {
 }
 
 pub fn backup_code_from_random() -> HashSet<String> {
-    let mut set = std::collections::HashSet::with_capacity(8);
-    for _ in 1..8 {
-        set.insert(readable_password_from_random());
-    }
-    set
+    (1..9)
```
Comment (Firstyear) [inline kanidmd/src/lib/utils.rs:42]: Does this generate 8 codes? Just checking.

**Review by Firstyear — CHANGES_REQUESTED** [review]

File: kanidmd/src/lib/idm/authsession.rs (hunk 21 above)
Comment (victorcwai) [reply kanidmd/src/lib/idm/authsession.rs:321]: Good catch. So when the delayed action is processed, I should retrieve the backup codes and remove `code_to_remove` from it, right?

File: kanidmd/src/lib/utils.rs (hunk 22 above)
Comment (victorcwai) [reply kanidmd/src/lib/utils.rs:42]: It inserts 8 codes. But on second thought, if the `readable_password_from_random()` returns the same string, then the HashSet won't have 8 codes.

Comment (victorcwai) [reply kanidmd/src/lib/utils.rs:42]: I will change it to insert until set.size() <= 8

File: kanidmd/src/lib/core/https.rs (hunk 2 above)
Comment (Firstyear) [reply kanidmd/src/lib/core/https.rs:1306]: In the future there is a plan to have multiple credentials per account, but it's not fully fleshed out, but we want the routes to reflect this possible change for now.

File: kanidmd/src/lib/utils.rs (hunk 22 above)
Comment (Firstyear) [reply kanidmd/src/lib/utils.rs:42]: Just make it 0..8, that's clearer .... Also readable_password_from_random returns a new random PW each call.

File: kanidmd/src/lib/idm/authsession.rs (hunk 21 above)
Comment (Firstyear) [reply kanidmd/src/lib/idm/authsession.rs:321]: Correct! :D

File: kanidmd/src/lib/credential/mod.rs (hunk 8 above)
Comment (Firstyear) [reply kanidmd/src/lib/credential/mod.rs:837]: Ahhh yes, your right, I misread :)

**Review by Firstyear — COMMENTED** [review]
Hope these comments help :)

File: kanidmd/src/lib/idm/authsession.rs [hunk 23]
```diff
@@ -1950,4 +1987,200 @@ mod tests {
...
+                &async_tx,
+                &webauthn,
+                Some(&pw_badlist_cache),
+                &hs512,
+            ) {
+                Ok(AuthState::Denied(msg)) => assert!(msg == BAD_PASSWORD_MSG),
+                _ => panic!(),
+            };
+        }
+        // Can't process BackupCodeRemoval without the server instance
```
Comment (Firstyear) [inline kanidmd/src/lib/idm/authsession.rs:2106]: That's fine, you just have to deque here and check that the removal message *was* sent.

**Review by Firstyear — APPROVED** [review]
This looks really good. Is there anything else still missing here?

**Comment by victorcwai** [comment]
Feel free to merge!

**Comment by Firstyear** [comment]
Sure, do you also mind making a ticket for the kanidm_tools to be extended to create and reset these too? Shouldn't be a long jump from here to add that to the CLI.
